trie-db/src/trie_codec.rs

```diff
@@ -0,0 +1,635 @@
+// Copyright 2019 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Compact encoding/decoding functions for partial Merkle-Patricia tries.
+//!
+//! A partial trie is a subset of the nodes in a complete trie, which can still be used to
+//! perform authenticated lookups on a subset of keys. A naive encoding is the set of encoded nodes
+//! in the partial trie. This, however, includes redundant hashes of other nodes in the partial
+//! trie which could be computed directly. The compact encoding strips out all hash child
+//! references to other nodes in the partial trie and replaces them with empty inline references,
+//! indicating that the child reference is omitted. The nodes are then ordered in pre-order
+//! traversal order so that the full nodes can be efficiently reconstructed recursively. Note that
+//! hash references to nodes not in the partial trie are left intact. The compact encoding can be
+//! expected to save roughly (n - 1) hashes in size where n is the number of nodes in the partial
+//! trie.
+
+use hash_db::HashDB;
+use crate::{
+	CError, ChildReference, DBValue, NibbleVec, NodeCodec, Result,
+	TrieHash, TrieError, TrieDB, TrieDBNodeIterator, TrieLayout,
+	nibble_ops::NIBBLE_LENGTH, node::{Node, NodeHandle, NodeHandlePlan, NodePlan, OwnedNode},
+};
+
+use	::core_::{
+	convert::TryInto,
+	marker::PhantomData,
+	result,
+};
+
+#[cfg(feature = "std")]
+use std::rc::Rc;
+#[cfg(not(feature = "std"))]
+use alloc::rc::Rc;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+struct EncoderStackEntry<C: NodeCodec> {
+	/// The prefix is the nibble path to the node in the trie.
+	prefix: NibbleVec,
+	node: Rc<OwnedNode<DBValue>>,
+	/// The next entry in the stack is a child of the preceding entry at this index. For branch
+	/// nodes, the index is in [0, NIBBLE_LENGTH] and for extension nodes, the index is in [0, 1].
+	child_index: usize,
+	/// Flags indicating whether each child is omitted in the encoded node.
+	omit_children: Vec<bool>,
+	/// The encoding of the subtrie nodes rooted at this entry, which is built up in
+	/// `encode_compact`.
+	encoded_nodes: Vec<Vec<u8>>,
```

Comment:
I am not sure why it is in Entry: shouldn't it be outside this struct ?
Like using a single VecDeque and avoid those multiple vecs instance.

Reply:
I figured out how to get rid of this.